tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/WindowAggExec.java

```diff
@@ -347,6 +350,372 @@ private void finalizeWindow() {
+      //   if comp == null, there is no order by and window frame is set as the entire partition
+      if (comp != null && windowFrameUnit == WindowSpec.WindowFrameUnit.RANGE) {
+        // move frame end point to the last row of the same order by value
+        if (sameEndRange == 0) {
+          sameEndRange = numOfTuplesWithTheSameKeyValue(frameEnd, comp, true);
+        }
+        sameEndRange --;
+        frameEnd += sameEndRange;
+
+        // move frame start point to the first row of the same order by value
+        if (sameStartRange == 0) {
+          sameStartRange = numOfTuplesWithTheSameKeyValue(frameStart, comp, true);
+          actualStart = frameStart;
+        }
+        sameStartRange --;
+        frameStart = actualStart;
+      }
+      // As the number of built-in window functions that support window frame is small,
+      // special treatment for each function seems to be reasonable
+      Tuple inTuple = getFunctionSpecificInput(funcName, frameStart, frameEnd);
+      Datum result = NullDatum.get();
+
+      if (inTuple != null) {
+        functions[functionIndex].merge(contexts[functionIndex], inTuple);
+        result = functions[functionIndex].terminate(contexts[functionIndex]);
+      }
+      Tuple outTuple = evaluatedTuples.get(i);
+      outTuple.put(nonFunctionColumnNum + functionIndex, result);
+    }
+  }
+
+  // it returns the number of Tuples with the same order by key value
+  // if only one Tuple for the given order by key exists, it returns 1
+  private int numOfTuplesWithTheSameKeyValue(int startOffset, BaseTupleComparator comp, boolean isForward) {
+    int numberOfTuples = 0;
+
+    Tuple inTuple = accumulatedInTuples.get(startOffset);
+    if (isForward) {
+      do {
+        numberOfTuples++;
+        startOffset++;
+      } while (startOffset < accumulatedInTuples.size() && comp.compare(accumulatedInTuples.get(startOffset), inTuple) == 0);
+    } else {      // backward direction
+      do {
+        numberOfTuples++;
+        startOffset--;
+      } while (startOffset >= 0 && comp.compare(accumulatedInTuples.get(startOffset), inTuple) == 0);
+    }
+
+    return numberOfTuples;
+  }
+
+  private Tuple getFunctionSpecificInput(String funcName, int dataStart, int dataEnd) {
+    if (funcName.equals("first_value")) {
+      // check the frame start is within the partition
+      if (dataStart <= dataEnd && dataStart >= 0 && dataStart < accumulatedInTuples.size()) {
+        return accumulatedInTuples.get(dataStart);
+      }
+    } else if (funcName.equals("last_value")) {
+      // check the frame end is within the partition
+      if (dataStart <= dataEnd && dataEnd >= 0 && dataEnd < accumulatedInTuples.size()) {
+        return accumulatedInTuples.get(dataEnd);
+      }
+    }
+    return null;
+  }
+
+  private void evaluateAggregationFunction(int functionIndex, LogicalWindowSpec.LogicalWindowFrame windowFrame,
```

Comment:
I don't want to push you. You can reserve this refactoring work as another jira if you want.